localstack-core/localstack/utils/backoff.py

```diff
@@ -0,0 +1,95 @@
+import random
+import time
+from dataclasses import dataclass
+
+
+@dataclass
+class ExponentialBackoff:
+    """
+    ExponentialBackoff implements exponential backoff with randomization.
+    The backoff period increases exponentially for each retry attempt, with
+    optional randomization within a defined range.
+
+    next_backoff() is calculated using the following formula:
+        ```
+        randomized_interval = random_between(retry_interval * (1 - randomization_factor), retry_interval * (1 + randomization_factor))
+        ```
+
+    For example, given:
+        `initial_interval` = 2
+        `randomization_factor` = 0.5
+        `multiplier` = 2
+
+    The next backoff will be between 1 and 3 seconds (2 * [0.5, 1.5]).
+    The following backoff will be between 2 and 6 seconds (4 * [0.5, 1.5]).
+
+    Note:
+        - `max_interval` caps the base interval, not the randomized value
+        - Returns 0 when `max_retries` or `max_time_elapsed` is exceeded
+        - The implementation is not thread-safe
+
+    Example sequence with defaults (initial_interval=0.5, randomization_factor=0.5, multiplier=1.5):
+
+    | Request # | Retry Interval (seconds) | Randomized Interval (seconds) |
+    |-----------|----------------------|----------------------------|
+    | 1         | 0.5                  | [0.25, 0.75]              |
+    | 2         | 0.75                 | [0.375, 1.125]            |
+    | 3         | 1.125                | [0.562, 1.687]            |
+    | 4         | 1.687                | [0.8435, 2.53]            |
+    | 5         | 2.53                 | [1.265, 3.795]            |
+    | 6         | 3.795                | [1.897, 5.692]            |
+    | 7         | 5.692                | [2.846, 8.538]            |
+    | 8         | 8.538                | [4.269, 12.807]           |
+    | 9         | 12.807               | [6.403, 19.210]           |
+    | 10        | 19.210               | 0                         |
+
+    Note: The sequence stops at request #10 when `max_retries` or `max_time_elapsed` is exceeded
+    """
+
+    initial_interval: float = 0.5  # Initial backoff interval in seconds
+    randomization_factor: float = 0.5  # Factor to randomize backoff [0,1]
+    multiplier: float = 1.5  # Multiply interval by this factor each retry
+    max_interval: float = 60.0  # Maximum backoff interval in seconds
+    max_retries: int = -1  # Max retry attempts (-1 for unlimited)
+    max_time_elapsed: float = -1  # Max total time in seconds (-1 for unlimited)
+
+    def __post_init__(self):
+        self.retry_interval: float = 0
+        self.retries: int = 0
+        self.start_t: float = 0.0
```

Comment:
nit: even though the meaning is quite evident here, I'd suggest to not use magic letters in names. 
```suggestion
        self.start_time: float = 0.0
```